src/main/java/org/springframework/data/jpa/repository/query/DeclaredQuery.java

```diff
@@ -80,9 +82,12 @@ static DeclaredQuery of(@Nullable String query) {
 	 * 
 	 * @param countQuery an optional query string to be used if present.
 	 * @param countQueryProjection an optional return type for the query.
+	 * @param metadata metadata about the queried entity.
+	 * @param parser SpEL parser for evaluating expressions within the query.
 	 * @return a new {@literal DeclaredQuery} instance.
 	 */
-	DeclaredQuery deriveCountQuery(@Nullable String countQuery, @Nullable String countQueryProjection);
+	DeclaredQuery deriveCountQuery(@Nullable String countQuery, @Nullable String countQueryProjection,
```

Comment:
Adding the parser as an argument here makes this rather basic class dependent on infrastructure which is really only needed for `ExpressionBasedStringQuery`. I'd therefor rather leave this method unchanged and instead add a factory method in `ExpressionBasedStringQuery` which takes a `DeclaredQuery` and turns it into an `ExpressionBasedStringQuery`. That method can then be used in `AbstractStringBasedJpaQuery`.